File: LeetCode/W1/linkedListCycle.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function hasCycle(head: ListNode | null): boolean {
    const nodeSet = new Set<ListNode | null>()
    let result: boolean = false

    while (head && !result) {
      {nodeSet.has(head)
        ? 
        result = true 
        : 
        nodeSet.add(head)
        head = head.next}
    }
    return result;
};

File: LeetCode/phase1/BinarySearch/LinkedListCycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {

    public boolean hasCycle(ListNode head) {

      //base
      if (head == null || head.next == null) {
        return false;
      }

      //define pointers
      ListNode slow = head, fast = head;

      //solution
      while (fast != null && fast.next != null) {
        
        slow = slow.next;
        fast = fast.next.next;

        if (slow == fast) {
          return true;
        }

      }
    
      return false;
    }


}

//Input: head = [3,2,0,-4]